operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,1743 @@
+                            column,
+                            features.num_features(),
+                            true,
+                        )
+                    }
+                    _ => todo!(), // strings, categories
+                }?;
+            }
+
+            // TODO: time buffer
+
+            self.feature_output_buffers.push(FeatureOutputBuffers {
+                geo: geo_buffers,
+                numbers,
+                decimals,
+            })
+        }
+
+        Ok(())
+    }
+
+    fn set_feature_column_input_argument<T: OclPrm>(
+        kernel: &Kernel,
+        idx: usize,
+        column: &str,
+        len: usize,
+        data: &[T],
+        nulls: Option<&[bool]>,
+    ) -> Result<()> {
+        let buffer = Buffer::<T>::builder()
+            .queue(kernel.default_queue().expect("expect").clone())
+            .len(len)
+            .copy_host_slice(data)
+            .build()?;
+        kernel.set_arg(format!("IN_POINT{}_COLUMN_{}", idx, column), &buffer)?;
+
+        if let Some(nulls) = nulls {
+            // TODO: convert booleans to i32 for host device transfer more efficiently
+            let nulls: Vec<i32> = nulls.iter().map(|n| n.as_()).collect();
+            let buffer = Buffer::<i32>::builder()
+                .queue(kernel.default_queue().expect("expect").clone())
+                .len(len)
+                .copy_host_slice(nulls.as_slice())
+                .build()?;
+            kernel.set_arg(format!("IN_POINT{}_NULLS_{}", idx, column), &buffer)?;
+        }
+
+        Ok(())
+    }
+
+    fn set_feature_column_output_argument<T: OclPrm>(
+        buffers: &mut Vec<ColumnBuffer<T>>,
+        kernel: &Kernel,
+        idx: usize,
+        column: &str,
+        len: usize,
+        nullable: bool,
+    ) -> Result<()> {
+        let values = Buffer::<T>::builder()
+            .queue(kernel.default_queue().expect("expect").clone())
+            .len(len)
+            .build()?;
+        kernel.set_arg(format!("OUT_POINT{}_COLUMN_{}", idx, column), &values)?;
+
+        let nulls = if nullable {
+            let buffer = Buffer::<i32>::builder()
+                .queue(kernel.default_queue().expect("expect").clone())
+                .len(len)
+                .build()?;
+            kernel.set_arg(format!("OUT_POINT{}_NULLS_{}", idx, column), &buffer)?;
+
+            Some(buffer)
+        } else {
+            None
+        };
+
+        buffers.push(ColumnBuffer {
+            column_name: column.to_owned(),
+            values,
+            nulls,
+        });
+        Ok(())
+    }
+
+    fn set_raster_arguments(&mut self, kernel: &Kernel) -> Result<()> {
+        ensure!(
+            self.input_rasters.iter().all(Option::is_some),
+            error::CLProgramUnspecifiedRaster
+        );
+
+        for (idx, raster) in self.input_rasters.iter().enumerate() {
+            let raster = raster.expect("checked");
+            call_generic_raster2d!(raster, raster => {
```

Comment:
why use a macro here? i think this is something a trait method could handle very well..

Reply:
I need to create a `Buffer<T>` of the correct type and access to the rasters data container. I am open to suggestions on how to accept rasters as traits with a mixture of different T's.

Reply:
well you could use a Trait with a method into_buffer and then implement it for Raster<T> and the TypedRaster. Then you can just call it and you will get the correct type.

Reply:
but then I would need a large match block in the impl for `TypedRaster`. We explicitly added the macro to avoid the match blocks.

Reply:
I tried the trait method but it led to much more code and still couldn't avoid the match-macro. We should leave it as it is for now